### PR TITLE
Grammar fixes thanks to the friend who actually speaks english (:

### PR DESCRIPTION
**Pull request by bwplotka**
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: web/content/posts/golang-memory-monitoring.md [hunk 1]
```diff
@@ -13,115 +13,113 @@ categories:
...
 
-Comparing to the previous result, you need to admit that the first impression for Go 1.12.5 is quite negative, right? Memory usage is getting super high as memory stays high and never comes back. 
-Initially, it can look like a leak or some bug in memory allocation. Even worse it looks like the container is just about to OOM.
+Compared to the result with Go 1.11 you need to admit that the first impression for Go 1.12.5 is quite negative, right? Memory usage is increasing without ever coming down. 
+Initially, this might look like a leak or some bug in memory allocation. Even worse it looks like the container is just about to run out of memory (OOM).
 
-All those assumptions are wrong. In my case, the container is not even close to OOM (Out of memory signal), and it is actually more performant ([in theory](https://github.com/golang/go/issues/23687#issuecomment-496705293)). 
+When taking a close look, all those assumptions are actually proven wrong. In my case, the container is not even close to OOM, and it is actually more performant ([in theory](https://github.com/golang/go/issues/23687#issuecomment-496705293)).
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: a closer look

File: web/content/posts/golang-memory-monitoring.md [hunk 2]
```diff
@@ -13,115 +13,113 @@ categories:
...
+### Go 1.12 memory optimizations affect the reported memory usage / RSS in certain cases. What changed and why?
 
-The change responsible for this effect is roughly explained in [runtime](https://golang.org/doc/go1.12#runtime) release notes:
+The change responsible for this effect is roughly explained in the release notes for the [Go runtime](https://golang.org/doc/go1.12#runtime)::
 
 > On Linux, the runtime now uses MADV_FREE to release unused memory. This is more efficient but may result in higher reported RSS. The kernel will reclaim the unused data when it is needed. To revert to the Go 1.11 behavior (MADV_DONTNEED), set the environment variable GODEBUG=madvdontneed=1.
 
-As you know Go has quite sophisticated [GC mechanism](https://blog.golang.org/ismmkeynote0) which is responsible for freeing allocated memory whenever internal data structure is not used.
+As you probably know Go has quite sophisticated [GC mechanisms](https://blog.golang.org/ismmkeynote0) which are responsible for freeing allocated memory whenever a piece of data is no longer references by your code.
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: no longer referenced

File: web/content/posts/golang-memory-monitoring.md [hunk 3]
```diff
@@ -13,115 +13,113 @@ categories:
...
 The key part for our issue is that processes can release allocated memory in many different ways. 
-Those ways also depend on OS and kernel versions.  
-Among many options Go runtime in certain cases uses [`madvise`](http://man7.org/linux/man-pages/man2/madvise.2.html) system call. One of the many advantages
-of `madvise` is that Go process can **cooperate** with kernel closely on how to treat certain "pages" of the RAM memory in virtual space in a way that helps both sides.
+Those ways also vary per OS and kernel versions.
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: kernel version.

File: web/content/posts/golang-memory-monitoring.md [hunk 4]
```diff
@@ -13,115 +13,113 @@ categories:
...
+Those ways also vary per OS and kernel versions.  
+Among many options the Go runtime in certain cases uses [`madvise`](http://man7.org/linux/man-pages/man2/madvise.2.html) system call. One of the many advantages
+of `madvise` is that Go processes can closely **cooperate** with the kernel on how to treat certain "pages" of RAM memory in virtual space in a way that helps both sides.
 
 This cooperation is mutually beneficial, as programs generally use memory in a highly dynamic way. Sometimes they allocate more, sometimes less. 
 Since asking the kernel for free memory pages is sometimes quite expensive, doing that back and forth can take time and resources.
 On the other hand, we cannot just keep the memory reserved, as it can lead to the machine OOMing (kernel panic) or swapping to disk (extremely slow)
-if suddenly other processed requires it. Memory is non-compressible resource, so we need something in the middle: releasing by advising.
+if suddenly other process require more memory. Memory is a non-compressible resource so we need something in the middle: releasing by advising.
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: processes require

File: web/content/posts/golang-memory-monitoring.md [hunk 5]
```diff
@@ -13,115 +13,113 @@ categories:
...
-I don't think that's the professional name for this, but let's name it "cached memory pages" for the purpose of this blog.
+I don't think that it is the professional name for this approach, but let's call it "cached memory pages" for the purpose of this blog.
 
-**This approach affects the counter of memory occupied by a process in kernel. It's because those pages that are "cached" are still technically reserved by Go process, 
-even though kernel will use it as soon as it needs memory for other processes**
+**This approach affects the amount of memory occupied by a process as registered by the  kernel. Those "cached" pages are still technically reserved for the Go process, 
+even though the kernel can use them as soon as it needs memory for other processes**
 
-In details `madvise` system call in high level consists of 3 arguments:
+From high-level the `madvise` system call consists of 3 arguments:
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: From a high-level perspective the

File: web/content/posts/golang-memory-monitoring.md [hunk 6]
```diff
@@ -13,115 +13,113 @@ categories:
...
 
-* `address` and `length` that defines what memory range this call refers to.
+* `address` and `length` that define what memory range this call refers to.
 * `advice` that says what to advice for those memory pages.
 
-`advice` can have many different values depending on OS and kernel version.
+`advice` can have many different values depending on the specific OS and kernel version used by the system on which the Go process is running.
 
-To explain Go 1.12 change, we are interested in those two:
+To explain the Go 1.12 change, we are interested in following two values:
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: interested in two specific values:

File: web/content/posts/golang-memory-monitoring.md [hunk 7]
```diff
@@ -13,115 +13,113 @@ categories:
...
      specified by addr and len.  The kernel can thus free these
      pages, but the freeing could be delayed until memory pressure
      occurs.  (...)
 
-In the essence Go 1.11 was based mostly on `MADV_DONTNEED` whereas Go 1.12 if possible uses `MADV_FREE`. As you can read in descriptions, the latter
-tells the kernel to not free resources associated with given range until memory pressure occurs. Memory pressure means 
-other process or kernel itself has not enough memory in the unused pool.
+In essence Go 1.11 was mostly using `MADV_DONTNEED` whereas Go 1.12 where possible uses `MADV_FREE`. As you can read in the descriptions above, the latter
+tells the kernel to not free resources associated with the given range until memory pressure occurs. Memory pressure means 
+that other processes or the kernel itself does not have enough memory in the unused pool to satisfy their needs.
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: itself do not have

File: web/content/posts/golang-memory-monitoring.md [hunk 8]
```diff
@@ -13,115 +13,113 @@ categories:
...
 
-In my opinion, this change makes a lot of sense, especially in Kubernetes/container environment. When you think about it, the pattern is to use a single process per container. 
-Singe the memory limits are per container as well it means that releasing memory immediately by the only process that is running on the container is mostly wasted work.
+In my opinion, this change makes a lot of sense, especially in a Kubernetes/container environment, where the general pattern is to use a single process per container. 
+Since memory limits are enforced on a per container basic, releasing memory immediately from the only process that is running inside of it is mostly wasted work.
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: per container basis

Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: releasing memory for the only

File: web/content/posts/golang-memory-monitoring.md [hunk 9]
```diff
@@ -13,115 +13,113 @@ categories:
...
-* Application perspective: How much we allocate and where (heap vs stack etc).
+* The application perspective: How much we do allocate and where (heap vs stack etc).
 
-Here we are quite lucky as Go gives a handful of metrics. With Prometheus client enabled they all look like this (during the same test as in the beginning:
+Here we are quite lucky as Go gives a handful of metrics. With the Prometheus client enabled they all look like this (during the same test as at the beginning of this blog post):
 
-![in-use memory of Prometheus from Go perspective during test](/images/blog/go-memory-monitoring/3.png)
+![in-use memory of Prometheus from the Go runtime's perspective during test](/images/blog/go-memory-monitoring/3.png)
 
 All those metrics are 1:1 fetched from [runtime.MemStats](https://golang.org/pkg/runtime/#MemStats)
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: are fetched without alterations from

File: web/content/posts/golang-memory-monitoring.md [hunk 10]
```diff
@@ -13,115 +13,113 @@ categories:
...
 
-This is more tricky. Let's focus on the container here. Cadvisor exposes container's memory metrics straight from [cgroup memory controller](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt) counters 
+This is more tricky. Let's focus on the container here. Cadvisor exposes a container's memory metrics straight from the [cgroup memory controller](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt) counters 
 ([code](https://sourcegraph.com/github.com/google/cadvisor@cc445b9cc7e20e12062cc40ac0aa2b88c40dc487/-/blob/container/libcontainer/handler.go#L533))
 
-As you remember `container_memory_usage_bytes` was not useful very well. Essentially you never know if memory is saturated or just cached.
-Even worse `usage_in_bytes` cgroup counter was always quite fuzzy, non-exact counter, from [cgroup doc](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt):
+As you remember `container_memory_usage_bytes` was not very useful. Essentially you never know if memory is saturated or just cached.
+Even worse, the `usage_in_bytes` cgroup counter is quite fuzzy and non-exact. From the [cgroup docs](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt):
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: counter is quite approximative.

File: web/content/posts/golang-memory-monitoring.md [hunk 11]
```diff
@@ -132,12 +130,12 @@ Even worse `usage_in_bytes` cgroup counter was always quite fuzzy, non-exact cou
   
 There is `container_memory_rss` but it behaves similarly due to `MADV_FREE` behavior.
 
-The only promising metric is `container_memory_working_set_bytes` recommended on various channels. It generally behaves similarly too `container_memory_usage_bytes`:
+The only promising metric is `container_memory_working_set_bytes` recommended by various sources. It generally behaves similarly to `container_memory_usage_bytes`:
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: Nit: would be good to list some of those sources.

File: web/content/posts/golang-memory-monitoring.md [hunk 12]
```diff
@@ -179,34 +177,35 @@ corev1.ResourceRequirements{
...
-* Do not afraid to update Go runtime version in your application. But when you do:
+* If you depend on `container_memory_usage_bytes` switch to `container_memory_working_set_bytes` metric for closest possible experience to actual usage. It's not perfect though.
+* Use `go_memstats_alloc_bytes` and others (e.g `go_memstats_.*_inuse_bytes`) to see actual allocations. Useful when profiling and optimizing your application memory. This helps to filter out the memory that is "cached" and it's the most accurate from the application perspective.
+* Avoid `Go 1.12.0-1.12.4` due to a memory [allocation slowness bug](https://github.com/kubernetes/kubernetes/issues/75833#issuecomment-487830238) 
+* Do not be afraid to update the Go runtime version in your application. But when you do:
   * Read the changelog
-  * Change JUST the version (: Change single thing at the time to ensure that if there is something suspicious, you can immediately narrow to Go runtime upgrade.  
+  * Change JUST the version (: Change a single thing at a time to ensure that if there is something suspicious, you can immediately narrow down on the Go runtime upgrade.
```
Comment (Helcaraxan) [inline web/content/posts/golang-memory-monitoring.md]: narrow down to the

**Comment by bwplotka** [comment]
Addressed in https://github.com/bwplotka/my/commit/e6705572ab30d769d0fa1b18fb4c2e2a7e86df84

Thanks!

**Pull request merged**